### ISSUE TITLE
Implement "allOf" of object types as single class

### ISSUE THREAD

**Issue by schani** [quicktype/quicktype#541]
This is a stop-gap measure before we have subtypes or interfaces.

* If all elements in an `allOf` are object types that allow additional properties, we can make a single class out of them.

* If a property is required in any one of the elements then it's non-optional.

* Properties that are object types need to be allOf-ified recursively if they occur in more than one element.

* The same goes for the item types of arrays and maps.

* Integers and numbers become integers.

* Enums and strings become the intersection of all the enums.

**Comment by schani** [comment]
Here's where we are after #574:

Intersections types, i.e. `allOf`, are now resolved to more basic types.  To do that it was necessary to change how we process unions from JSON Schema as well, since unions and intersections can be nested.  The way we do this is that we have two passes, one for flattening unions, the other for resolving intersections, which run alternating, until there is nothing left to do.

There is at least one problem left to fix.  Consider this type:

```
AllOf(
    {"foo": string},
    Union(
        {"bar": string},
        Map<string>
    )
)
```

Flattening the union results in `Map<string>`, i.e., we lose any information on the `bar` property. Resolving the intersection then gives just `{"foo": string}`, but what we actually want is `{"foo": string, "bar": string}`.

I’ll have to combine the class and map types into one, like JSON Schema does, where one can specify individual properties, and then a type for additional properties.  That combined type will live until all the unions and intersections are done, then it'll be change into either a pure class or map type, so the renderers won't be affected.

We might at some later point want to support that kind of type in the renderers, too.

**Comment by schani** [comment]
Implemented in #696
